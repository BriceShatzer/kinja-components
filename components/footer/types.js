/* @flow */

export type { Blog } from '../types';
export type { AnalyticsInjectedProps } from '../hoc/analytics';
