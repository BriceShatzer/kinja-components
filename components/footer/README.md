# Footer :shoe: ![ready](status-images/ready.svg)

The footer that is rendered at the bottom of a page on Kinja. Also includes logic for the store and subscription modals triggered by links in the footer.

[Storybook Demo](http://localhost:8001/?selectedKind=Footer).

<!-- STORY -->

##### Table of Contents
 - [Subcomponents](#subcomponents)
 - [Props](#props)
   - [blog](#props-blog)
   - [feature](#props-feature)
   - [isPermalink](#props-permalink)
   - [language](#props-language)
   - [showStoreDialog](#props-store)
 - [Example](#example)

<a name="subcomponents"/>

### Subcomponents

 - [Blog Logo](./subcomponents/footerLogo.js)
 - [General Links](./subcomponents/footerLinks.js)
 - [Legal Disclaimers](./subcomponents/footerLegal.js)
 - [Network Links](./subcomponents/footerNetwork.js)
 - [Social Links](./subcomponents/footerSocial.js)

#### Modals
 - [Store Redirect](./modals/storeRedirect.js)

<a name="props"/>

### Props

<a name="props-blog"/>

#### blog

###### type: _BlogType_

The blog corresponding to the page that the footer is being rendered on. If it exists, the footer will try to render the blog's logo and links to its about and store pages. Otherwise it will render a generic footer.

<a name="props-feature"/>

#### feature

###### type: _FeatureType [required]_

The feature switch controller to determine if various switches are on or off.

<a name="props-permalink"/>

#### isPermalink

###### type: _boolean_, default: false

Whether the footer is being rendered on a permalink page or not.

<a name="props-language"/>

#### language

###### type: _String_, default: 'en-US'

The language to render translatable messages to. Only affects links in the `FooterLinks` component.

<a name="props-store"/>

#### showStoreDialog

###### type: _boolean_

Optional override to show the store modal when the component is first rendered.

<a name="example"/>

### Example

```javascript
const { blog } = window.kinja.meta;

<Footer
	blog={blog}
	feature={feature}
	language={'es'}
/>
```
