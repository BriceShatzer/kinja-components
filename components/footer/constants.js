/* @flow */

export const legalDisclaimers = {
	theonion: 'The Onion is not intended for readers under 18 years of age.',
	clickhole: `ClickHole uses invented names in all of its stories, except in cases where public figures
							are being satirized. Any other use of real names is accidental and coincidental. ClickHole is
							not intended for readers under 18 years of age.`
};

export const leavingSiteDisclaimer = (legalText: string) => (
	`Disclaimer: You are leaving ${legalText} website and going to a third party site,
	 which is subject to its own privacy policy and terms of use.`
);

export const onionIncBlogs = [
	'avclub',
	'clickhole',
	'theonion',
	'thetakeout'
];

export const gmgBlogs = [
	'deadspin',
	'earther',
	'gizmodo',
	'jalopnik',
	'jezebel',
	'kotaku',
	'lifehacker',
	'splinter',
	'theinventory',
	'theroot'
];

export const storeUrls = {
	jalopnik: 'https://store.jalopnik.com',
	gizmodo: 'https://store.gizmodo.com',
	kotaku: 'https://store.kotaku.com',
	jezebel: 'https://store.jezebel.com',
	lifehacker: 'https://store.lifehacker.com',
	deadspin: 'https://store.deadspin.com',
	avclub: 'https://store.avclub.com/',
	theonion: 'https://store.theonion.com/',
	theroot: 'https://store.theroot.com/',
	thetakeout: 'https://store.thetakeout.com/'
};
