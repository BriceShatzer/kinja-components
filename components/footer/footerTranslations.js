/* @flow */

export default {
	'es-ES': {
		'Need Help?': '¿Necesitas ayuda?',
		'Content Guide': 'Pautas de Contenido',
		'Privacy': 'Privacidad',
		'Terms of Use': 'Términos de uso',
		'Advertising': 'Publicidad',
		'Jobs': 'Empleo',
		'Sitemap': 'Mapa del sitio'
	}
};
