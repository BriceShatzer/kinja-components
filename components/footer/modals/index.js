/* @flow */

import StoreRedirect from './storeRedirect';

export {
	StoreRedirect
};
