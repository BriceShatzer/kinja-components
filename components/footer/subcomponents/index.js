/* @flow */

import FooterLegal from './footerLegal';
import FooterLinks from './footerLinks';
import FooterLogo from './footerLogo';
import FooterNetwork from './footerNetwork';
import FooterSocial from './footerSocial';
import { FooterList } from './footerLinks';

export {
	FooterLegal,
	FooterLinks,
	FooterLogo,
	FooterNetwork,
	FooterSocial,
	FooterList
};
