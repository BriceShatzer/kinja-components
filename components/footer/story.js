/* @flow */

import * as React from 'react';
import styled from 'styled-components';
import {
	action,
	boolean,
	select,
	storiesOf,
	withDocs
} from 'base-storybook';

import Footer from './footer';
import {
	FooterLegal,
	FooterLinks,
	FooterLogo,
	FooterNetwork,
	FooterSocial
} from './subcomponents';
import { StoreRedirect } from './modals';

import footerTranslations from './footerTranslations';
import Modal from '../modal';
import createTranslate from '../translator';
import { NETWORK_BLOGS } from '../../config/consts';
import README from './README.md';

import blog1 from '../../__stubs__/gizmodo.json';
import blog2 from '../../__stubs__/theonion.json';
import blog3 from '../../__stubs__/avclub.json';
import Blog from 'kinja-magma/models/Blog';

const blogs = {
	[blog1.displayName]: Blog.fromJSON(blog1),
	[blog2.displayName]: Blog.fromJSON(blog2),
	[blog3.displayName]: Blog.fromJSON(blog3)
};

const blogOptions = Object.keys(blogs);

const CenteredGreyDiv = styled.div`
	align-items: center;
	background: #222;
	display: flex;
	justify-content: space-around;
`;

const wrapInFooter = child => (
	<CenteredGreyDiv>
		{child}
	</CenteredGreyDiv>
);

const FullWidthWrapper = styled.div`
	flex: 1;
`;

storiesOf('4. Components|Navigation/Footer', module)
	.addDecorator(withDocs(README))
	.add('Blog Footer', () => {
		const blog = blogs[select('blog', blogOptions, blogOptions[0])];
		const isNonBlogPage = boolean('Non-blog page', false);
		const showFooterLogo = boolean('Show Footer Logo', true);
		const enableStoreLink = boolean('Enable store link', true);

		const language = select('Locale', {
			'en-US': 'English (en)',
			'es-ES': 'Spanish (es)'
		}, 'en-US');

		const footerProps = {
			blog: isNonBlogPage ? undefined : blog,
			feature: { isOn: () => true },
			ga: (s1: mixed, s2: mixed, s3: mixed) => action('ga event')(...[s1, s2, s3]),
			language,
			showFooterLogo: !isNonBlogPage && showFooterLogo,
			enableStoreLink,
			enableSitemap: false
		};

		return (
			<FullWidthWrapper>
				<Footer {...footerProps} />
			</FullWidthWrapper>
		);
	})
	.add('Generic Footer', () => {
		const language = select('Locale', {
			'en-US': 'English (en)',
			'es-ES': 'Spanish (es)'
		}, 'en-US');

		const footerProps = {
			feature: { isOn: () => true },
			ga: (s1: mixed, s2: mixed, s3: mixed) => action('ga event')(...[s1, s2, s3]),
			language,
			enableSitemap: false
		};

		return (
			<FullWidthWrapper>
				<Footer {...footerProps} redirectHandler={(url: string) => action('redirect')(url)}/>
			</FullWidthWrapper>
		);
	})
	.add('Legal', () => {
		const blog = blogs[select('blog', blogOptions, blogOptions[0])];

		const legalProps = {
			blogGroup: blog.blogGroup
		};

		return wrapInFooter(
			<FooterLegal {...legalProps} />
		);
	})
	.add('Links', () => {
		const translate = createTranslate(footerTranslations, select('Locale', {
			'en-US': 'English (en)',
			'es-ES': 'Spanish (es)'
		}, 'en-US'));

		const blog = blogs[select('blog', blogOptions, blogOptions[0])];

		const linksProps = {
			showAboutLink: Boolean(blog.properties.aboutPostId),
			feature: { isOn: () => true },
			ga: (s1: mixed, s2: mixed, s3: mixed) => action('ga event')(...[s1, s2, s3]),
			toggleStoreModal: () => action('opening store modal!'),
			translate,
			enableSitemap: false
		};

		return wrapInFooter(
			<FooterLinks {...linksProps} />
		);
	})
	.add('Logo', () => {
		const blog = blogs[select('blog', blogOptions, blogOptions[0])];

		const logoProps = {
			blogGroup: blog.blogGroup
		};

		return wrapInFooter(
			<FooterLogo {...logoProps} />
		);
	})
	.add('Network', () => {
		const blog = blogs[select('blog', blogOptions, blogOptions[0])];

		const networkProps = {
			blogGroup: blog.blogGroup,
			networkBlogs: NETWORK_BLOGS,
			ga: (s1: mixed, s2: mixed, s3: mixed) => action('ga event')(...[s1, s2, s3])
		};

		return wrapInFooter(
			<FooterNetwork {...networkProps} />
		);
	})
	.add('Social', () => {
		const blog = blogs[select('blog', blogOptions, blogOptions[0])];

		const {
			facebookScreenName,
			twitterScreenName,
			instagramScreenName,
			youtubeUrl,
			canonicalHost
		} = blog;

		const socialProps = {
			canonicalHost,
			facebookScreenName,
			ga: (s1: mixed, s2: mixed, s3: mixed) => action('ga event')(...[s1, s2, s3]),
			instagramScreenName,
			twitterScreenName,
			youtubeUrl
		};

		return wrapInFooter(
			<FooterSocial {...socialProps} />
		);
	})
	.add('Store Redirect Modal', () => {
		const blog = blogs[select('blog', blogOptions, blogOptions[0])];

		const modalProps = {
			blur: true,
			fullscreen: true,
			isOpen: boolean('Is open', true),
			onClose: () => action('modal')('close')
		};

		const storeRedirectProps = {
			blogGroup: blog.blogGroup,
			displayName: blog.displayName,
			redirectHandler: (url: string) => action('redirect')(url),
			toggleStoreModal: () => action('modal')('close')
		};

		return (
			<Modal {...modalProps} >
				<StoreRedirect {...storeRedirectProps} />
			</Modal>
		);
	});
