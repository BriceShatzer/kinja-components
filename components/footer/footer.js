/* @flow */

import * as React from 'react';
import styled from 'styled-components';

import { BLOG_GROUPS } from '../../config/consts';
import footerTranslations from './footerTranslations';
import Analytics from '../hoc/analytics';
import type { AnalyticsInjectedProps }  from '../hoc/analytics';
import OnScreen from '../hoc/on-screen';
import type { OnScreenInjectedProps } from '../hoc/on-screen';
import {
	FooterLegal,
	FooterLinks,
	FooterLogo,
	FooterNetwork,
	FooterSocial
} from './subcomponents';
import createTranslate from '../translator';
import type { TranslateFunction } from '../translator';
import { EnsureDefaultTheme } from '../theme';

import { FooterList } from './subcomponents';
import Link from '../elements/link';

import type Blog from 'kinja-magma/models/Blog';

type FooterProps = {
	blog?: ?Blog,
	enableStoreLink?: boolean,
	enableSourcePoint?: boolean,
	enableSitemap: boolean,
	hideContentGuide?: boolean,
	language?: string,
	showFooterLogo?: boolean
} & AnalyticsInjectedProps & OnScreenInjectedProps;

const FooterWrap = styled.footer`
	padding-bottom: 0;
	width: 100%;
`;

const FooterNav = styled.div`
	background-color: ${props => props.theme.color.darksmoke};
	padding-top: 40px;

	@media and (max-width: ${props => props.theme.breakpointMedium}) {
		overflow-x: hidden;
	}
`;

const SourcePointListWrapper = styled.div`
	padding-bottom: 38px;
`;
const SourcePointList = styled(FooterList)``;

export class Footer extends React.Component<FooterProps> {
	static displayName = 'Footer';

	static defaultProps = {
		language: 'en-US'
	};

	translate: TranslateFunction;

	constructor(props: FooterProps) {
		super(props);

		this.translate = createTranslate(footerTranslations, props.language);
	}

	componentWillReceiveProps = (newProps: FooterProps) => {
		if (this.props.language !== newProps.language) {
			this.translate = createTranslate(footerTranslations, newProps.language);
		}
		if (!this.props.isVisible && newProps.isVisible) {
			this.props.ga('Footer', 'scroll', 'complete', {nonInteraction: 1});
		}
	}

	renderGenericFooter() {
		const footerLinksProps = {
			ga: this.props.ga,
			translate: this.translate,
			enableSitemap: false
		};

		return (
			<EnsureDefaultTheme>
				<FooterWrap className="js_footer">
					<FooterNav>
						<FooterLinks {...footerLinksProps} />
						<FooterNetwork ga={this.props.ga} />
						<FooterLegal />
					</FooterNav>
				</FooterWrap>
			</EnsureDefaultTheme>
		);
	}

	renderBlogFooter(blog: Blog) {
		const { enableStoreLink, ga, hideContentGuide, showFooterLogo, enableSourcePoint, enableSitemap } = this.props;
		const { blogGroup } = blog;
		const displayName = BLOG_GROUPS[blogGroup] || blog.displayName;
		const showAboutLink = Boolean(blog.properties && blog.properties.aboutPostId);
		const footerLinksProps = {
			enableStoreLink,
			enableSitemap,
			blogGroup,
			displayName,
			hideContentGuide,
			showAboutLink,
			ga,
			translate: this.translate
		};

		const footerNetworkProps = {
			blogGroup,
			ga
		};

		const footerSocialProps = {
			canonicalHost: blog.canonicalHost,
			facebookScreenName: blog.facebookScreenName,
			ga,
			instagramScreenName: blog.instagramScreenName,
			showFooterLogo,
			twitterScreenName: blog.twitterScreenName,
			youtubeUrl: blog.youtubeUrl
		};

		const shouldRenderConsentLink = enableSourcePoint && window.__consent__ !== undefined;
		const gdprMessage = this.translate('Manage Privacy Settings');
		const ccpaMessage = this.translate('Do Not Sell My Data');

		const renderGdprLink = () => {
			return <SourcePointList>
				<li>
					<Link onClick={() => window._sp_.loadPrivacyManagerModal(6626, '5e2889d2f2436f01dc3c1135')}>
						{gdprMessage}
					</Link>
				</li>
			</SourcePointList>;
		};

		const renderCcpaLink = () => {
			return <SourcePointList>
				<li>
					<Link onClick={() => window._sp_ccpa.loadPrivacyManagerModal(null, '5e1e327bf97533429a96040b')}>
						{ccpaMessage}
					</Link>
				</li>
			</SourcePointList>;
		};

		const sourcePointLinks = () => {
			const isGdpr = shouldRenderConsentLink && window.__consent__ === 'GDPR';
			const isCcpa = shouldRenderConsentLink && window.__consent__ === 'CCPA';
			if (isGdpr) {
				return renderGdprLink();
			} else if (isCcpa) {
				return renderCcpaLink();
			}
		};

		return (
			<EnsureDefaultTheme>
				<FooterWrap className="js_footer">
					<FooterNav>
						{showFooterLogo ? <FooterLogo blogGroup={blogGroup} /> : <div className="js_footer-logo"/>}
						<FooterSocial {...footerSocialProps} />
						<FooterLinks {...footerLinksProps} />
						{shouldRenderConsentLink &&
							<SourcePointListWrapper>
								{sourcePointLinks()}
							</SourcePointListWrapper>
						}
						<FooterNetwork {...footerNetworkProps} />
						<FooterLegal blogGroup={blogGroup} />
					</FooterNav>
				</FooterWrap>
			</EnsureDefaultTheme>
		);
	}

	render() {
		const { blog } = this.props;

		return blog && blog.name ? this.renderBlogFooter(blog) : this.renderGenericFooter();
	}
}

export default OnScreen(Analytics(Footer));
