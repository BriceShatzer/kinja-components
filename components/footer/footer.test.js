import * as React from 'react';
import { shallow } from 'enzyme';

import { Footer } from './footer';

describe('<Footer />', () => {
	const blog = {
		aboutPostId: '',
		blogGroup: '',
		canonicalHost: '',
		displayName: 'display name',
		facebookScreenName: '',
		id: '',
		instagramScreenName: '',
		isCommerce: false,
		isGmgBlog: false,
		language: '',
		locale: '',
		name: 'test',
		parentId: '',
		properties: {
			aboutPostId: ''
		},
		timezone: '',
		twitterScreenName: '',
		youtubeUrl: ''
	};

	const stubbedProps = {
		ga: jest.fn(),
		isPermalink: false,
		showFooterLogo: false
	};

	test('should render generic footer w/o blog', () => {
		const wrapper = shallow(<Footer { ...stubbedProps } />);

		expect(wrapper).toMatchSnapshot();
	});

	test('should render blog footer w/ blog', () => {
		const wrapper = shallow(<Footer { ...stubbedProps } blog={ blog } />);

		expect(wrapper).toMatchSnapshot();
	});

	test('should not show footer logo by default', () => {
		const wrapper = shallow(<Footer { ...stubbedProps } blog={ blog } />);

		expect(wrapper.find('FooterLogo').exists()).toBe(false);
	});

	test('should render placeholder when lazy loading enabled', () => {
		const wrapper = shallow(<Footer { ...stubbedProps } blog={ blog } showFooterLogo />);

		expect(wrapper.find('FooterWrap')).toMatchSnapshot();
	});
});
