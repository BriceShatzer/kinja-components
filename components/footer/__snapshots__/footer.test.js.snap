// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> should render blog footer w/ blog 1`] = `
<EnsureDefaultTheme>
  <footer__FooterWrap
    className="js_footer"
  >
    <footer__FooterNav>
      <div
        className="js_footer-logo"
      />
      <FooterSocial
        canonicalHost=""
        facebookScreenName=""
        ga={[MockFunction]}
        instagramScreenName=""
        showFooterLogo={false}
        twitterScreenName=""
        youtubeUrl=""
      />
      <FooterLinks
        blogGroup=""
        displayName="display name"
        ga={[MockFunction]}
        showAboutLink={false}
        translate={[Function]}
      />
      <FooterNetwork
        blogGroup=""
        ga={[MockFunction]}
      />
      <FooterLegal
        blogGroup=""
      />
    </footer__FooterNav>
  </footer__FooterWrap>
</EnsureDefaultTheme>
`;

exports[`<Footer /> should render generic footer w/o blog 1`] = `
<EnsureDefaultTheme>
  <footer__FooterWrap
    className="js_footer"
  >
    <footer__FooterNav>
      <FooterLinks
        enableSitemap={false}
        ga={[MockFunction]}
        translate={[Function]}
      />
      <FooterNetwork
        ga={[MockFunction]}
      />
      <FooterLegal />
    </footer__FooterNav>
  </footer__FooterWrap>
</EnsureDefaultTheme>
`;

exports[`<Footer /> should render placeholder when lazy loading enabled 1`] = `null`;
